### DISCUSSION TITLE
How do I re-use a widget?

### DISCUSSION THREAD

**Discussion by tronje** [Textualize/textual#5230] [Q&A] [answered]
I want to show an "action log" in my app that's visible on all screens. So the user can navigate a tree of menu screens, but each screen should show the same instance of `RichLog`, which keeps a log of actions the user has taken.

I created something like this:

```python
class GlobalLog(RichLog):
    _instance = None

    @classmethod
    def instance(cls):
        if GlobalLog._instance is None:
            GlobalLog._instance = cls(highlight=True, markup=True)

        return GlobalLog._instance
```

and I `yield GlobalLog.instance()` from the `compose` methods of my screens. But it only ever shows up on the first screen, Subsequent screens do not show this widget at all, and there's also no error. I'm guessing perhaps because widgets are assigned IDs and those must be unique in the DOM? Maybe?

Any help is appreciated, thanks!

**Comment by willmcgugan** [comment]
You can't re-use widgets. You will need to construct an instance when you add it to the DOM.

Having a rich log showing the same data on all screens is an odd requirements. I suspect this may be an [XY problem](https://label.dev/articles/xy-problem/). Maybe if you explain what you are trying to solve, I can offer suggestions.

**Reply by tronje** [reply]
I see, thanks.

I feel like I explained what I want to do fairly well, but let me try again. I have an app with a tree-like menu structure. Each menu is shown as a different screen, with different menu entries.
But I want to show a persistent log on each screen, that shows the actions the user has previously taken. For that, I felt like `RichLog` is a good choice.

For the sake of example, say it's an app that controls smart lights. One menu allows setting lights on/off, one menu allows setting the color, one allows assigning names to lights or whatever.
If the user does these actions:
* set light A on
* set light B off
* set light A's color to purple
* rename light C to D

I want all screens in my app to display this log of actions.

**Reply by tronje** [reply]
I suppose one solution would be to keep this log outside of any `textual` object, and just re-play it for every screen that is created, with a new instance of `RichLog`, but that feels inefficient.

**Reply by willmcgugan** [reply]
For reference, the following is your proposed *solution* to your problem:

"but each screen should show the same instance of RichLog"

Your *problem* would be something like this:

"How can I have a log on multiple screens, showing the same data?"

If your log is a list of text, then I'd recommend storing it in a list, and writing it to your log on ScreenResume. It will do some additional work, but unless your log is extremely large performance won't be an issue.

**Reply by tronje** [reply, marked as answer]
> For reference, the following is your proposed _solution_ to your problem:
> 
> "but each screen should show the same instance of RichLog"
> 
> Your _problem_ would be something like this:
> 
> "How can I have a log on multiple screens, showing the same data?"

... sure

> If your log is a list of text, then I'd recommend storing it in a list, and writing it to your log on ScreenResume. It will do some additional work, but unless your log is extremely large performance won't be an issue.

I have a global list of log entries now, and whenever I append a log, I add it to the active screen's log. The `on_screen_resume` handler clears the log and then fills it again with all log entries, to avoid duplicates. Thanks, that solves it.
